// @alitajs/dform/lib/components/RangeDatePicker/demo/datetime.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 时间区间选择框
 * desc: 日期和时间
 */
var DateTime = function DateTime() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "rangeTime1",
    fieldProps2: "rangeTime2",
    title: "\u65F6\u95F4(datetime)",
    modeType: "datetime",
    required: true,
    firstProps: {
      onOk: function onOk(val) {
        // eslint-disable-next-line no-console
        console.log(val);
      }
    },
    secondProps: {
      onOk: function onOk(val) {
        // eslint-disable-next-line no-console
        console.log(val);
      }
    }
  })));
};

var _default = DateTime;
exports.default = _default;
