// @alitajs/dform/lib/components/ExtraInput/demo/positionInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _position_ico = _interopRequireDefault(require("../../../assets/position_ico.png"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 复杂输入框
 * desc: 位置输入框
 */
// eslint-disable-next-line no-console
var extraImg = function extraImg() {
  return /*#__PURE__*/_react.default.createElement("img", {
    src: _position_ico.default,
    onClick: function onClick(e) {
      return console.log(e);
    }
  });
};

var PositionInput = function PositionInput() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "minPosition",
    fieldProps2: "maxPosition",
    title: "\u4F4D\u7F6E\u533A\u95F4",
    placeholder: "\u9009\u62E9\u6700\u5C0F\u4F4D\u7F6E",
    placeholder2: "\u9009\u62E9\u6700\u5927\u4F4D\u7F6E",
    required: true,
    inputType: "text",
    extra: extraImg()
  })));
};

var _default = PositionInput;
exports.default = _default;
