// @alitajs/dform/lib/components/ExtraInput/demo/priceInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 复杂输入框
 * desc: 价格输入框
 */
var PriceInput = function PriceInput() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "minPrise",
    fieldProps2: "maxPrise",
    title: "\u4EF7\u683C\u533A\u95F4(\u6570\u5B57\u8F93\u5165)",
    placeholder: "\u8F93\u5165\u6700\u5C0F\u4EF7\u683C",
    placeholder2: "\u8F93\u5165\u6700\u5927\u4EF7\u683C",
    required: true,
    inputType: "number",
    labelNumber: 8,
    extra: "\xA5"
  })));
};

var _default = PriceInput;
exports.default = _default;
