// @alitajs/dform/lib/components/ExtraInput/demo/selectExtra.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 复杂输入框
 * desc: 选择单位，输入价格
 */
var unitList = [[{
  label: '元',
  value: '元'
}, {
  label: '万元',
  value: '万元'
}, {
  label: '亿元',
  value: '亿元'
}]];

var SelectExtra = function SelectExtra() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "price",
    fieldProps2: "unit",
    title: "\u5355\u4EF7",
    placeholder: "\u8F93\u5165\u4EF7\u683C",
    placeholder2: "\u9009\u62E9\u533A\u95F4",
    required: true,
    extraType: "select",
    data: unitList
  })));
};

var _default = SelectExtra;
exports.default = _default;
