// @alitajs/dform/lib/components/ExtraInput/demo/verticalExtraInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 复杂输入框
 * desc: 垂直样式，输入长度区间
 */
var VerticalExtraInput = function VerticalExtraInput() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "minLength",
    fieldProps2: "maxLength",
    title: "\u957F\u5EA6\u533A\u95F4",
    placeholder: "\u8F93\u5165\u957F\u5EA6",
    placeholder2: "\u8F93\u5165\u957F\u5EA6",
    positionType: "vertical",
    required: true
  })));
};

var _default = VerticalExtraInput;
exports.default = _default;
