// @alitajs/dform/lib/components/ExtraInput/demo/verticalExtraInputAndSelect.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 复杂输入框
 * desc: 垂直样式，输入框和选择框
 */
var unitList = [[{
  label: '元',
  value: '元'
}, {
  label: '万元',
  value: '万元'
}, {
  label: '亿元',
  value: '亿元'
}]];

var VerticalExtraInputAndSelect = function VerticalExtraInputAndSelect() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    title: "\u4EF7\u683C",
    fieldProps: "prices",
    fieldProps2: "priceUnit",
    placeholder: "\u8F93\u5165\u957F\u5EA6",
    placeholder2: "\u9009\u62E9\u957F\u5EA6\u5355\u4F4D",
    positionType: "vertical",
    extraType: "select",
    data: unitList
  })));
};

var _default = VerticalExtraInputAndSelect;
exports.default = _default;
