// @alitajs/dform/lib/components/NomarRadio/demo/verticalRadioType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 Radio
 * desc: 垂直按钮
 */
var foodList = [{
  label: '宫保鸡丁',
  value: '宫保鸡丁'
}, {
  label: '可乐鸡翅',
  value: '可乐鸡翅'
}, {
  label: '爆炒虾仁',
  value: '爆炒虾仁'
}, {
  label: '清蒸小黄鱼',
  value: '清蒸小黄鱼'
}, {
  label: '红烧肉',
  value: '红烧肉'
}];

var VerticalRadioType = function VerticalRadioType() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "userRadio5",
    required: true,
    data: foodList,
    title: "\u559C\u6B22\u7684\u98DF\u7269",
    positionType: "vertical",
    radioType: "vertical"
  })));
};

var _default = VerticalRadioType;
exports.default = _default;
