// @alitajs/dform/lib/components/NomarRadio/demo/verticalRadio.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 Radio
 * desc: 垂直按钮
 */
var dayList = [{
  label: '晴',
  value: '晴'
}, {
  label: '阴',
  value: '阴'
}, {
  label: '雨',
  value: '雨'
}];

var VerticalRadio = function VerticalRadio() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "userRadio4",
    required: true,
    data: dayList,
    title: "\u5929\u6C14\u60C5\u51B5",
    positionType: "vertical"
  })));
};

var _default = VerticalRadio;
exports.default = _default;
