// @alitajs/dform/lib/components/NomarTextArea/demo/extraArea.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

var _photo = _interopRequireDefault(require("../../../assets/photo.png"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 多行文本输入框
 * desc: 右边注释
 */
var photoImg = function photoImg() {
  return /*#__PURE__*/_react.default.createElement("img", {
    src: _photo.default,
    style: {
      width: '3rem',
      height: '2rem'
    }
  });
};

var ExtraArea = function ExtraArea() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "idenPhone",
    title: "\u8EAB\u4EFD\u8BC1",
    placeholder: "\u5B58\u5728 extra \u81EA\u52A8\u6362\u884C",
    required: true,
    extra: photoImg()
  })));
};

var _default = ExtraArea;
exports.default = _default;
