// @alitajs/dform/lib/components/NomarTextArea/demo/titleTooLong.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 多行文本输入框
 * desc: 标题文字内容过长
 */
var NomarArea = function NomarArea() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "titleTooLong",
    required: true,
    placeholder: "\u8BF7\u8F93\u5165...",
    title: "\u6807\u9898\u6587\u5B57\u5185\u5BB9\u8FC7\u957F",
    labelNumber: 8
  })));
};

var _default = NomarArea;
exports.default = _default;
