// @alitajs/dform/lib/components/NomarInput/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/input-item/style");

var _inputItem = _interopRequireDefault(require("antd-mobile/es/input-item"));

var _react = _interopRequireDefault(require("react"));

var _Field = _interopRequireDefault(require("../Field"));

require("../../styles/index.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var NomarInput = function NomarInput(props) {
  var _props$inputType = props.inputType,
      inputType = _props$inputType === void 0 ? 'text' : _props$inputType,
      coverStyle = props.coverStyle,
      _props$title = props.title,
      title = _props$title === void 0 ? '' : _props$title,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      fieldProps = props.fieldProps,
      rules = props.rules,
      _props$positionType = props.positionType,
      positionType = _props$positionType === void 0 ? 'horizontal' : _props$positionType,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      extra = props.extra,
      subTitle = props.subTitle,
      _props$hidden = props.hidden,
      hidden = _props$hidden === void 0 ? false : _props$hidden,
      onBlur = props.onBlur,
      otherProps = _objectWithoutProperties(props, ["inputType", "coverStyle", "title", "required", "fieldProps", "rules", "positionType", "hasStar", "extra", "subTitle", "hidden", "onBlur"]);

  var isVertical = positionType === 'vertical';

  var inputOnBlur = function inputOnBlur(val) {
    window.scrollTo(0, 0);
    if (onBlur) onBlur(val);
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !hidden && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-input-title"
  }, isVertical && /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-vertical-title"
  }, required && hasStar && /*#__PURE__*/_react.default.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title), subTitle), extra !== '' && isVertical && /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-area-extra"
  }, extra)), /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform".concat(isVertical ? '-vertical' : '', "-input")
  }, /*#__PURE__*/_react.default.createElement(_Field.default, {
    name: fieldProps,
    rules: rules || [{
      required: required,
      message: "\u8BF7\u8F93\u5165".concat(title)
    }]
  }, /*#__PURE__*/_react.default.createElement(_inputItem.default, _extends({}, otherProps, {
    extra: isVertical ? '' : extra,
    type: inputType,
    style: _objectSpread({
      textAlign: isVertical ? 'left' : 'right'
    }, coverStyle),
    onBlur: function onBlur(val) {
      inputOnBlur(val);
    }
  }), required && hasStar && /*#__PURE__*/_react.default.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title))))));
};

var _default = NomarInput;
exports.default = _default;
