// @alitajs/dform/lib/components/NomarPicker/demo/onlyRead.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 Select
 * desc: 只读，不可编辑
 */
var citys = [[{
  label: '福州',
  value: '福州'
}, {
  label: '厦门',
  value: '厦门'
}]];

var OnlyRead = function OnlyRead() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "userPicker3",
    required: true,
    title: "\u57CE\u5E02(\u4E0D\u53EF\u7F16\u8F91)",
    placeholder: "\u8BF7\u9009\u62E9",
    data: citys,
    disabled: true
  })));
};

var _default = OnlyRead;
exports.default = _default;
