// @alitajs/tabs-layout/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

var _getLayoutContent = _interopRequireDefault(require("./utils/getLayoutContent"));

var _getTabsLayout = _interopRequireDefault(require("./utils/getTabsLayout"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DIR_NAME = 'tabs-layout';
const MODEL_NAME = 'Tabs';
const RELATIVE_MODEL = (0, _path().join)(DIR_NAME, MODEL_NAME);

var _default = api => {
  if (api.userConfig.keepalive) {
    console.error('请不要和 keep alive 插件一起使用，两个功能相同');
    return;
  }

  ;

  if (!api.userConfig.tabsLayout) {
    return;
  }

  ;
  api.describe({
    key: 'tabsLayout',
    config: {
      default: [/./],

      schema(joi) {
        return joi.array();
      },

      onChange: api.ConfigChangeType.regenerateTmpFiles
    }
  });
  api.onGenerateFiles(() => {
    api.writeTmpFile({
      path: (0, _path().join)(DIR_NAME, 'TabsLayout.tsx'),
      content: (0, _getTabsLayout.default)(api.paths.absTmpPath || '')
    });
    api.writeTmpFile({
      path: (0, _path().join)(DIR_NAME, 'Tabs.tsx'),
      content: (0, _getLayoutContent.default)(api.userConfig.tabsLayout, './TabsLayout')
    });
  });
  api.addUmiExports(() => [{
    exportAll: true,
    source: `../${RELATIVE_MODEL}`
  }]);
};

exports.default = _default;
