// @alitajs/dform/lib/components/NomarInput/demo/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/button/style");

var _button = _interopRequireDefault(require("antd-mobile/es/button"));

require("antd-mobile/es/white-space/style");

var _whiteSpace = _interopRequireDefault(require("antd-mobile/es/white-space"));

var _react = _interopRequireWildcard(require("react"));

var _dform = _interopRequireWildcard(require("@alitajs/dform"));

var _position_ico = _interopRequireDefault(require("../../../assets/position_ico.png"));

var _photo = _interopRequireDefault(require("../../../assets/photo.png"));

var _look = _interopRequireDefault(require("../../../assets/look.png"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var Page = function Page() {
  var _useForm = (0, _dform.useForm)(),
      _useForm2 = _slicedToArray(_useForm, 1),
      form = _useForm2[0];

  var _useState = (0, _react.useState)(true),
      _useState2 = _slicedToArray(_useState, 2),
      pwdInputType = _useState2[0],
      setPwdInputType = _useState2[1];

  var onFinish = function onFinish(values) {
    // eslint-disable-next-line no-console
    console.log('Success:', values);
  };

  var onFinishFailed = function onFinishFailed(errorInfo) {
    // eslint-disable-next-line no-console
    console.log('Failed:', errorInfo);
  };

  var extraImg = function extraImg() {
    return /*#__PURE__*/_react.default.createElement("img", {
      src: _position_ico.default
    });
  };

  var photoImg = function photoImg() {
    return /*#__PURE__*/_react.default.createElement("img", {
      src: _photo.default,
      style: {
        width: '3rem',
        height: '2rem'
      }
    });
  };

  var subTitle = function subTitle() {
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        color: 'red'
      }
    }, "\u6B64\u4E3A\u5FC5\u586B\u9879(\u526F\u6807\u9898)");
  };

  var pwdImg = function pwdImg() {
    return /*#__PURE__*/_react.default.createElement("img", {
      src: _look.default,
      onClick: function onClick() {
        setPwdInputType(!pwdInputType);
      }
    });
  };

  var formsData = [{
    type: 'input',
    fieldProps: 'username',
    required: true,
    placeholder: '输入项居左',
    title: '用户名',
    inputType: 'text',
    clear: true,
    subTitle: subTitle(),
    coverStyle: {
      textAlign: 'left'
    }
  }, {
    type: 'input',
    fieldProps: 'userAge',
    required: true,
    placeholder: '请输入',
    title: '年龄',
    editable: false,
    inputType: 'text'
  }, {
    type: 'input',
    fieldProps: 'userPwd',
    required: true,
    title: '请设置密码',
    extra: pwdImg(),
    inputType: pwdInputType ? 'password' : 'text'
  }, {
    type: 'input',
    fieldProps: 'userPrice',
    required: true,
    placeholder: '0.00',
    title: '价格',
    extra: '¥',
    inputType: 'number'
  }, {
    type: 'input',
    fieldProps: 'userPosition',
    required: true,
    placeholder: '请定位',
    title: '定位',
    extra: extraImg(),
    inputType: 'number'
  }, {
    type: 'input',
    fieldProps: 'username5',
    required: true,
    placeholder: '',
    title: '身份证',
    extra: photoImg(),
    inputType: 'number'
  }, {
    type: 'input',
    fieldProps: 'userTitle',
    required: true,
    placeholder: '点击获取表单全部数据',
    title: '标题',
    editable: false,
    onClick: function onClick() {
      // eslint-disable-next-line no-console
      console.log(form.getFieldsValue());
    }
  }, {
    type: 'input',
    fieldProps: 'titleTooLong',
    required: true,
    placeholder: '请输入',
    title: '标题名称过长',
    labelNumber: 7,
    inputType: 'text',
    clear: true
  }, {
    type: 'input',
    fieldProps: 'titleTooLong2',
    required: true,
    placeholder: '请输入',
    title: '标题名称过长(超过14个字符自动换行)',
    inputType: 'text',
    extra: photoImg(),
    clear: true
  }, {
    type: 'input',
    fieldProps: 'cardNumber',
    required: true,
    placeholder: '请输入',
    title: '身份证号码(增加规则)',
    inputType: 'text',
    clear: true,
    subTitle: subTitle(),
    positionType: 'vertical',
    rules: [{
      required: true,
      message: "\u8BF7\u8F93\u5165"
    }, {
      pattern: new RegExp(/^[0-9a-zA-Z_]{1,}$/, 'g'),
      message: '名称只允许包含数字、字母和下划线'
    }]
  }];
  var formsValues = {
    userAge: '这里只读不可编辑',
    username4: '点击图标事件',
    username6: '存在点击事件',
    userTitle: '点击获取表单全部数据',
    userPwd: '123456'
  };
  var formProps = {
    onFinish: onFinish,
    onFinishFailed: onFinishFailed,
    data: formsData,
    formsValues: formsValues,
    form: form,
    isDev: true
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_dform.default, formProps), /*#__PURE__*/_react.default.createElement(_whiteSpace.default, {
    size: "sm"
  }), /*#__PURE__*/_react.default.createElement(_button.default, {
    type: "primary",
    onClick: function onClick() {
      return form.submit();
    }
  }, "Submit"));
};

var _default = Page;
exports.default = _default;
