// @alitajs/dform/lib/components/NomarInput/demo/nodeExtraClick.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _position_ico = _interopRequireDefault(require("../../../assets/position_ico.png"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 输入框
 * desc: extra 可以自定义样式，并且可设置点击事件
 */
var NodeExtraClick = function NodeExtraClick() {
  var extraImg = function extraImg() {
    return /*#__PURE__*/_react.default.createElement("img", {
      src: _position_ico.default
    });
  };

  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "userPosition",
    required: true,
    placeholder: "\u8BF7\u5B9A\u4F4D",
    title: "\u5B9A\u4F4D",
    editable: false,
    extra: extraImg()
  })));
};

var _default = NodeExtraClick;
exports.default = _default;
