// @alitajs/dform/lib/components/NomarInput/demo/textClick.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 输入框
 * desc: 存在文本点击事件
 */
var TextClick = function TextClick() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "userTitle",
    required: true,
    placeholder: "\u5B58\u5728\u70B9\u51FB\u4E8B\u4EF6",
    title: "\u6807\u9898" // eslint-disable-next-line no-console
    ,
    onClick: function onClick(e) {
      return console.log(e);
    },
    editable: false
  })));
};

var _default = TextClick;
exports.default = _default;
