// @alitajs/dform/lib/components/NomarInput/demo/titleTooLong.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 输入框
 * desc: 标题名称过长
 */
var TitleTooLong = function TitleTooLong() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "usernameTooLong",
    required: true,
    placeholder: "\u8BF7\u8F93\u5165",
    title: "\u6807\u9898\u540D\u79F0\u8FC7\u957F",
    inputType: "text",
    labelNumber: 7,
    clear: true
  })));
};

var _default = TitleTooLong;
exports.default = _default;
