// @alitajs/dform/lib/components/NomarInput/demo/verticalInput.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireDefault(require("react"));

var _rcFieldForm = _interopRequireDefault(require("rc-field-form"));

var _ = _interopRequireDefault(require(".."));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * title: 基础 输入框
 * desc: 垂直方向的输入框
 */
var VerticalInput = function VerticalInput() {
  return /*#__PURE__*/_react.default.createElement(_rcFieldForm.default, null, /*#__PURE__*/_react.default.createElement(_list.default, null, /*#__PURE__*/_react.default.createElement(_.default, {
    fieldProps: "cardNumber",
    required: true,
    placeholder: "\u8BF7\u8F93\u5165",
    title: "\u8EAB\u4EFD\u8BC1\u53F7\u7801",
    inputType: "text",
    clear: true,
    positionType: "vertical",
    rules: [{
      required: true,
      message: "\u8BF7\u8F93\u5165"
    }, {
      pattern: new RegExp(/^[0-9a-zA-Z_]{1,}$/, 'g'),
      message: '名称只允许包含数字、字母和下划线'
    }]
  })));
};

var _default = VerticalInput;
exports.default = _default;
