// @alitajs/dform/es/components/MultiplePicker/multiplePickerGroup.js
import "antd-mobile/es/modal/style";
import _Modal from "antd-mobile/es/modal";
import "antd-mobile/es/list/style";
import _List from "antd-mobile/es/list";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useState, useEffect } from 'react';
import difference from 'lodash/difference';
import classnames from 'classnames';
import { InputItem } from '..';
import '../../styles/index.less';
var Item = _List.Item;

var MultiplePickerGroup = function MultiplePickerGroup(props) {
  var _props$data = props.data,
      data = _props$data === void 0 ? [] : _props$data,
      title = props.title,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? '请选择' : _props$placeholder,
      onChange = props.onChange,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      _props$positionType = props.positionType,
      positionType = _props$positionType === void 0 ? 'horizontal' : _props$positionType,
      _props$initValue = props.initValue,
      initValue = _props$initValue === void 0 ? [] : _props$initValue,
      maxValueLength = props.maxValueLength,
      coverStyle = props.coverStyle,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      fieldProps = props.fieldProps,
      _props$labelNumber = props.labelNumber,
      labelNumber = _props$labelNumber === void 0 ? 5 : _props$labelNumber,
      _onClick = props.onClick;

  var _useState = useState([]),
      _useState2 = _slicedToArray(_useState, 2),
      context = _useState2[0],
      setContext = _useState2[1];

  var _useState3 = useState([]),
      _useState4 = _slicedToArray(_useState3, 2),
      preContext = _useState4[0],
      setPreContext = _useState4[1];

  var _useState5 = useState([]),
      _useState6 = _slicedToArray(_useState5, 2),
      preInitValue = _useState6[0],
      setPreInitValue = _useState6[1];

  var _useState7 = useState([]),
      _useState8 = _slicedToArray(_useState7, 2),
      selList = _useState8[0],
      setSelList = _useState8[1];

  var _useState9 = useState(false),
      _useState10 = _slicedToArray(_useState9, 2),
      modalFlag = _useState10[0],
      setModalFlag = _useState10[1];

  var _useState11 = useState(''),
      _useState12 = _slicedToArray(_useState11, 2),
      multipleLabel = _useState12[0],
      setMultipleLabel = _useState12[1];

  var isVertical = positionType === 'vertical';
  useEffect(function () {
    if (!data || data.length === 0) return;

    if (context.length === 0 || difference(initValue, preInitValue).length !== 0) {
      var dataList = JSON.parse(JSON.stringify(data));
      var selLabelList = [];
      var selValueList = [];
      setContext(_toConsumableArray(dataList).map(function (item) {
        var initItem = item;

        if (initValue.indexOf(initItem.value) !== -1) {
          if (!maxValueLength || maxValueLength && maxValueLength > selValueList.length) {
            initItem.flag = true;
            selLabelList.push(item.label);
            selValueList.push(item.value);
          } else {
            initItem.flag = false;
          }
        } else {
          initItem.flag = false;
        }

        return initItem;
      }));
      setMultipleLabel(selLabelList.join(','));
      setPreInitValue(initValue);
      setSelList(selValueList);
    }
  }, [data, initValue]);

  var pickerClick = function pickerClick(val) {
    var dataList = JSON.parse(JSON.stringify(context));

    if (selList.indexOf(val.value) !== -1) {
      selList.splice(selList.indexOf(val.value), 1);
    } else {
      selList.push(val.value);
    }

    if (maxValueLength && selList.length > maxValueLength) {
      selList.shift();
    }

    setContext(_toConsumableArray(dataList).map(function (item) {
      var initItem = item;

      if (selList.indexOf(initItem.value) !== -1) {
        initItem.flag = true;
      } else {
        initItem.flag = false;
      }

      return initItem;
    }));
  };

  var openMoal = function openMoal() {
    if (disabled) return;
    setPreContext(_toConsumableArray(context));
    setModalFlag(true);
  };

  var onCancel = function onCancel() {
    setModalFlag(false);
    setContext(_toConsumableArray(preContext));
  };

  var onConfirm = function onConfirm() {
    var selLabelList = context.filter(function (it) {
      return it.flag;
    }).map(function (item) {
      return item.label;
    });
    var selValueList = context.filter(function (it) {
      return it.flag;
    }).map(function (item) {
      return item.value;
    });
    setMultipleLabel(selLabelList.join(','));
    setModalFlag(false);
    if (onChange) onChange(selValueList);
  };

  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(InputItem, {
    isVertical: isVertical,
    value: multipleLabel,
    placeholder: placeholder,
    labelNumber: labelNumber,
    coverStyle: coverStyle,
    onClick: function onClick() {
      if (_onClick) _onClick();
      openMoal();
    },
    readOnly: true
  }, required && hasStar && /*#__PURE__*/React.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/React.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title)), /*#__PURE__*/React.createElement(_Modal, {
    popup: true,
    visible: modalFlag,
    onClose: function onClose() {
      onCancel();
    },
    className: "alitajs-dform-multiple-picker",
    animationType: "slide-up",
    title: /*#__PURE__*/React.createElement("div", {
      className: "am-picker-popup-header"
    }, /*#__PURE__*/React.createElement("div", {
      className: "am-picker-popup-item am-picker-popup-header-left",
      onClick: function onClick() {
        onCancel();
      }
    }, "\u53D6\u6D88"), /*#__PURE__*/React.createElement("div", {
      className: "am-picker-popup-item am-picker-popup-title"
    }, title), /*#__PURE__*/React.createElement("div", {
      className: "am-picker-popup-item am-picker-popup-header-right",
      onClick: function onClick() {
        onConfirm();
      }
    }, "\u786E\u5B9A"))
  }, /*#__PURE__*/React.createElement(_List, {
    className: "alitajs-dform-modal-content"
  }, _toConsumableArray(context).map(function (item) {
    return /*#__PURE__*/React.createElement(Item, {
      key: item.value
    }, /*#__PURE__*/React.createElement("div", {
      className: "alitajs-dform-multiple-picker-item",
      onClick: function onClick() {
        pickerClick(item);
      }
    }, /*#__PURE__*/React.createElement("div", {
      className: classnames({
        'alitajs-dform-multiple-picker-label': true,
        'alitajs-dform-multiple-picker-checked': item.flag
      })
    }, item.label), item.flag && /*#__PURE__*/React.createElement("div", {
      className: "alitajs-dform-tick"
    })));
  }))));
};

export default MultiplePickerGroup;
