// @alitajs/dform/es/components/RangeDatePicker/index.js
import "antd-mobile/es/date-picker/style";
import _DatePicker from "antd-mobile/es/date-picker";
import "antd-mobile/es/list/style";
import _List from "antd-mobile/es/list";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

import React, { useState } from 'react';
import classnames from 'classnames';
import Field from '../Field';
import { changeDateFormat } from '../../utils';
import '../../styles/index.less';

var RangeDatePicker = function RangeDatePicker(props) {
  var _useState = useState(),
      _useState2 = _slicedToArray(_useState, 2),
      beginDate = _useState2[0],
      setBeginDate = _useState2[1];

  var _useState3 = useState(),
      _useState4 = _slicedToArray(_useState3, 2),
      endDate = _useState4[0],
      setEndDate = _useState4[1];

  var fieldProps = props.fieldProps,
      fieldProps2 = props.fieldProps2,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? '开始时间' : _props$placeholder,
      _props$placeholder2 = props.placeholder2,
      placeholder2 = _props$placeholder2 === void 0 ? '结束时间' : _props$placeholder2,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      _props$modeType = props.modeType,
      modeType = _props$modeType === void 0 ? 'date' : _props$modeType,
      _props$rules = props.rules,
      rules = _props$rules === void 0 ? [] : _props$rules,
      title = props.title,
      minDate = props.minDate,
      maxDate = props.maxDate,
      _props$positionType = props.positionType,
      positionType = _props$positionType === void 0 ? 'vertical' : _props$positionType,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      secondProps = props.secondProps,
      firstProps = props.firstProps,
      subTitle = props.subTitle,
      _props$hidden = props.hidden,
      hidden = _props$hidden === void 0 ? false : _props$hidden,
      otherProps = _objectWithoutProperties(props, ["fieldProps", "fieldProps2", "placeholder", "placeholder2", "required", "modeType", "rules", "title", "minDate", "maxDate", "positionType", "hasStar", "secondProps", "firstProps", "subTitle", "hidden"]);

  var isVertical = positionType === 'vertical';
  return /*#__PURE__*/React.createElement(React.Fragment, null, !hidden && /*#__PURE__*/React.createElement(React.Fragment, null, isVertical && /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-vertical-title"
  }, required && hasStar && /*#__PURE__*/React.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/React.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title), subTitle), /*#__PURE__*/React.createElement("div", {
    className: classnames({
      'alitajs-dform-range-horizontal': !isVertical,
      'alitajs-dform-range-date-picker': true
    }),
    style: {
      justifyContent: isVertical ? 'space-between' : 'center',
      width: isVertical ? '' : '100%'
    }
  }, /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-begin".concat(isVertical ? '-vertical' : '', "-picker")
  }, /*#__PURE__*/React.createElement(Field, {
    name: fieldProps,
    rules: rules || [{
      required: required,
      message: "\u8BF7\u9009\u62E9".concat(title)
    }],
    shouldUpdate: function shouldUpdate(prevValue, nextValue) {
      setBeginDate(nextValue && nextValue[fieldProps]);
      return prevValue !== nextValue;
    }
  }, /*#__PURE__*/React.createElement(_DatePicker, _extends({}, otherProps, firstProps, {
    mode: modeType,
    extra: placeholder,
    minDate: minDate,
    maxDate: endDate || maxDate,
    title: title,
    format: function format(value) {
      return changeDateFormat(value, modeType);
    },
    onChange: function onChange(e) {
      setBeginDate(e);
    }
  }), /*#__PURE__*/React.createElement(_List.Item, null, required && hasStar && /*#__PURE__*/React.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/React.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title), /*#__PURE__*/React.createElement("span", {
    id: fieldProps2
  }))))), /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-line"
  }, "~"), /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-end".concat(isVertical ? '-vertical' : '', "-picker")
  }, /*#__PURE__*/React.createElement(Field, {
    name: fieldProps2,
    rules: rules || [{
      required: required,
      message: "\u8BF7\u9009\u62E9".concat(title)
    }],
    shouldUpdate: function shouldUpdate(prevValue, nextValue) {
      setEndDate(nextValue && nextValue[fieldProps2]);
      return prevValue !== nextValue;
    }
  }, /*#__PURE__*/React.createElement(_DatePicker, _extends({}, otherProps, secondProps, {
    extra: placeholder2,
    mode: modeType,
    minDate: beginDate || minDate,
    maxDate: maxDate,
    title: title,
    format: function format(value) {
      return changeDateFormat(value, modeType);
    },
    onChange: function onChange(e) {
      setEndDate(e);
    }
  }), /*#__PURE__*/React.createElement(_List.Item, {
    arrow: "horizontal"
  })))))));
};

export default RangeDatePicker;
