// @alitajs/dform/es/components/NomarTextArea/demo/index.js
import "antd-mobile/es/button/style";
import _Button from "antd-mobile/es/button";
import "antd-mobile/es/white-space/style";
import _WhiteSpace from "antd-mobile/es/white-space";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * title: 基础 多行文本输入框
 * desc: 表单使用 demo
 */
import React from 'react';
import DynamicForm, { useForm } from '@alitajs/dform';
import PhotoIcon from '../../../assets/photo.png';

var Page = function Page() {
  var _useForm = useForm(),
      _useForm2 = _slicedToArray(_useForm, 1),
      form = _useForm2[0];

  var onFinish = function onFinish(values) {
    // eslint-disable-next-line no-console
    console.log('Success:', values);
  };

  var onFinishFailed = function onFinishFailed(errorInfo) {
    // eslint-disable-next-line no-console
    console.log('Failed:', errorInfo);
  };

  var photoImg = function photoImg() {
    return /*#__PURE__*/React.createElement("img", {
      src: PhotoIcon,
      style: {
        width: '3rem',
        height: '2rem'
      }
    });
  };

  var formsData = [{
    type: 'area',
    fieldProps: 'textArea1',
    required: true,
    placeholder: '请输入...',
    title: '公司简介',
    positionType: 'horizontal'
  }, {
    type: 'area',
    fieldProps: 'textArea2',
    title: '有标题',
    placeholder: '只读，不可编辑',
    rows: 3,
    editable: false
  }, {
    type: 'area',
    fieldProps: 'titleTooLong',
    title: '标题文字内容过长',
    placeholder: '请输入',
    labelNumber: 8,
    required: true,
    coverStyle: {
      border: '1px solid #108ee9',
      background: '#fff'
    }
  }, {
    type: 'area',
    fieldProps: 'Remarks',
    title: '备注',
    placeholder: '请输入',
    required: true
  }, {
    type: 'area',
    fieldProps: 'idenPhone',
    title: '身份证',
    extra: photoImg(),
    placeholder: '存在 extra 自动换行',
    required: true
  }];
  var formsValues = {};
  var formProps = {
    onFinish: onFinish,
    onFinishFailed: onFinishFailed,
    data: formsData,
    formsValues: formsValues,
    form: form,
    isDev: true
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DynamicForm, formProps), /*#__PURE__*/React.createElement(_WhiteSpace, {
    size: "sm"
  }), /*#__PURE__*/React.createElement(_Button, {
    type: "primary",
    onClick: function onClick() {
      return form.submit();
    }
  }, "Submit"));
};

export default Page;
