// @alitajs/dform/lib/components/MultiplePicker/multiplePickerGroup.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/modal/style");

var _modal = _interopRequireDefault(require("antd-mobile/es/modal"));

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireWildcard(require("react"));

var _difference = _interopRequireDefault(require("lodash/difference"));

var _classnames = _interopRequireDefault(require("classnames"));

var _ = require("..");

require("../../styles/index.less");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var Item = _list.default.Item;

var MultiplePickerGroup = function MultiplePickerGroup(props) {
  var _props$data = props.data,
      data = _props$data === void 0 ? [] : _props$data,
      title = props.title,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? '请选择' : _props$placeholder,
      onChange = props.onChange,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      _props$positionType = props.positionType,
      positionType = _props$positionType === void 0 ? 'horizontal' : _props$positionType,
      _props$initValue = props.initValue,
      initValue = _props$initValue === void 0 ? [] : _props$initValue,
      maxValueLength = props.maxValueLength,
      coverStyle = props.coverStyle,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      fieldProps = props.fieldProps,
      _props$labelNumber = props.labelNumber,
      labelNumber = _props$labelNumber === void 0 ? 5 : _props$labelNumber,
      _onClick = props.onClick;

  var _useState = (0, _react.useState)([]),
      _useState2 = _slicedToArray(_useState, 2),
      context = _useState2[0],
      setContext = _useState2[1];

  var _useState3 = (0, _react.useState)([]),
      _useState4 = _slicedToArray(_useState3, 2),
      preContext = _useState4[0],
      setPreContext = _useState4[1];

  var _useState5 = (0, _react.useState)([]),
      _useState6 = _slicedToArray(_useState5, 2),
      preInitValue = _useState6[0],
      setPreInitValue = _useState6[1];

  var _useState7 = (0, _react.useState)([]),
      _useState8 = _slicedToArray(_useState7, 2),
      selList = _useState8[0],
      setSelList = _useState8[1];

  var _useState9 = (0, _react.useState)(false),
      _useState10 = _slicedToArray(_useState9, 2),
      modalFlag = _useState10[0],
      setModalFlag = _useState10[1];

  var _useState11 = (0, _react.useState)(''),
      _useState12 = _slicedToArray(_useState11, 2),
      multipleLabel = _useState12[0],
      setMultipleLabel = _useState12[1];

  var isVertical = positionType === 'vertical';
  (0, _react.useEffect)(function () {
    if (!data || data.length === 0) return;

    if (context.length === 0 || (0, _difference.default)(initValue, preInitValue).length !== 0) {
      var dataList = JSON.parse(JSON.stringify(data));
      var selLabelList = [];
      var selValueList = [];
      setContext(_toConsumableArray(dataList).map(function (item) {
        var initItem = item;

        if (initValue.indexOf(initItem.value) !== -1) {
          if (!maxValueLength || maxValueLength && maxValueLength > selValueList.length) {
            initItem.flag = true;
            selLabelList.push(item.label);
            selValueList.push(item.value);
          } else {
            initItem.flag = false;
          }
        } else {
          initItem.flag = false;
        }

        return initItem;
      }));
      setMultipleLabel(selLabelList.join(','));
      setPreInitValue(initValue);
      setSelList(selValueList);
    }
  }, [data, initValue]);

  var pickerClick = function pickerClick(val) {
    var dataList = JSON.parse(JSON.stringify(context));

    if (selList.indexOf(val.value) !== -1) {
      selList.splice(selList.indexOf(val.value), 1);
    } else {
      selList.push(val.value);
    }

    if (maxValueLength && selList.length > maxValueLength) {
      selList.shift();
    }

    setContext(_toConsumableArray(dataList).map(function (item) {
      var initItem = item;

      if (selList.indexOf(initItem.value) !== -1) {
        initItem.flag = true;
      } else {
        initItem.flag = false;
      }

      return initItem;
    }));
  };

  var openMoal = function openMoal() {
    if (disabled) return;
    setPreContext(_toConsumableArray(context));
    setModalFlag(true);
  };

  var onCancel = function onCancel() {
    setModalFlag(false);
    setContext(_toConsumableArray(preContext));
  };

  var onConfirm = function onConfirm() {
    var selLabelList = context.filter(function (it) {
      return it.flag;
    }).map(function (item) {
      return item.label;
    });
    var selValueList = context.filter(function (it) {
      return it.flag;
    }).map(function (item) {
      return item.value;
    });
    setMultipleLabel(selLabelList.join(','));
    setModalFlag(false);
    if (onChange) onChange(selValueList);
  };

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_.InputItem, {
    isVertical: isVertical,
    value: multipleLabel,
    placeholder: placeholder,
    labelNumber: labelNumber,
    coverStyle: coverStyle,
    onClick: function onClick() {
      if (_onClick) _onClick();
      openMoal();
    },
    readOnly: true
  }, required && hasStar && /*#__PURE__*/_react.default.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title)), /*#__PURE__*/_react.default.createElement(_modal.default, {
    popup: true,
    visible: modalFlag,
    onClose: function onClose() {
      onCancel();
    },
    className: "alitajs-dform-multiple-picker",
    animationType: "slide-up",
    title: /*#__PURE__*/_react.default.createElement("div", {
      className: "am-picker-popup-header"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "am-picker-popup-item am-picker-popup-header-left",
      onClick: function onClick() {
        onCancel();
      }
    }, "\u53D6\u6D88"), /*#__PURE__*/_react.default.createElement("div", {
      className: "am-picker-popup-item am-picker-popup-title"
    }, title), /*#__PURE__*/_react.default.createElement("div", {
      className: "am-picker-popup-item am-picker-popup-header-right",
      onClick: function onClick() {
        onConfirm();
      }
    }, "\u786E\u5B9A"))
  }, /*#__PURE__*/_react.default.createElement(_list.default, {
    className: "alitajs-dform-modal-content"
  }, _toConsumableArray(context).map(function (item) {
    return /*#__PURE__*/_react.default.createElement(Item, {
      key: item.value
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "alitajs-dform-multiple-picker-item",
      onClick: function onClick() {
        pickerClick(item);
      }
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)({
        'alitajs-dform-multiple-picker-label': true,
        'alitajs-dform-multiple-picker-checked': item.flag
      })
    }, item.label), item.flag && /*#__PURE__*/_react.default.createElement("div", {
      className: "alitajs-dform-tick"
    })));
  }))));
};

var _default = MultiplePickerGroup;
exports.default = _default;
