// @alitajs/dform/lib/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "NomarInput", {
  enumerable: true,
  get: function get() {
    return _NomarInput.default;
  }
});
Object.defineProperty(exports, "NomarPicker", {
  enumerable: true,
  get: function get() {
    return _NomarPicker.default;
  }
});
Object.defineProperty(exports, "NomarSwitch", {
  enumerable: true,
  get: function get() {
    return _NomarSwitch.default;
  }
});
Object.defineProperty(exports, "NomarTextArea", {
  enumerable: true,
  get: function get() {
    return _NomarTextArea.default;
  }
});
Object.defineProperty(exports, "OnlyReadInput", {
  enumerable: true,
  get: function get() {
    return _OnlyReadInput.default;
  }
});
Object.defineProperty(exports, "NomarDatePicker", {
  enumerable: true,
  get: function get() {
    return _NomarDatePicker.default;
  }
});
Object.defineProperty(exports, "NomarRadio", {
  enumerable: true,
  get: function get() {
    return _NomarRadio.default;
  }
});
Object.defineProperty(exports, "RangeDatePicker", {
  enumerable: true,
  get: function get() {
    return _RangeDatePicker.default;
  }
});
Object.defineProperty(exports, "ExtraInput", {
  enumerable: true,
  get: function get() {
    return _ExtraInput.default;
  }
});
Object.defineProperty(exports, "NomarCheckBox", {
  enumerable: true,
  get: function get() {
    return _NomarCheckBox.default;
  }
});
Object.defineProperty(exports, "CoverRadio", {
  enumerable: true,
  get: function get() {
    return _CoverRadio.default;
  }
});
Object.defineProperty(exports, "NomarImagePicker", {
  enumerable: true,
  get: function get() {
    return _NomarImagePicker.default;
  }
});
Object.defineProperty(exports, "NomarCustom", {
  enumerable: true,
  get: function get() {
    return _NomarCustom.default;
  }
});
Object.defineProperty(exports, "MultiplePicker", {
  enumerable: true,
  get: function get() {
    return _MultiplePicker.default;
  }
});
Object.defineProperty(exports, "InputItem", {
  enumerable: true,
  get: function get() {
    return _InputItem.default;
  }
});
Object.defineProperty(exports, "AddressPicker", {
  enumerable: true,
  get: function get() {
    return _AddressPicker.default;
  }
});

var _NomarInput = _interopRequireDefault(require("./NomarInput"));

var _NomarPicker = _interopRequireDefault(require("./NomarPicker"));

var _NomarSwitch = _interopRequireDefault(require("./NomarSwitch"));

var _NomarTextArea = _interopRequireDefault(require("./NomarTextArea"));

var _OnlyReadInput = _interopRequireDefault(require("./OnlyReadInput"));

var _NomarDatePicker = _interopRequireDefault(require("./NomarDatePicker"));

var _NomarRadio = _interopRequireDefault(require("./NomarRadio"));

var _RangeDatePicker = _interopRequireDefault(require("./RangeDatePicker"));

var _ExtraInput = _interopRequireDefault(require("./ExtraInput"));

var _NomarCheckBox = _interopRequireDefault(require("./NomarCheckBox"));

var _CoverRadio = _interopRequireDefault(require("./CoverRadio"));

var _NomarImagePicker = _interopRequireDefault(require("./NomarImagePicker"));

var _NomarCustom = _interopRequireDefault(require("./NomarCustom"));

var _MultiplePicker = _interopRequireDefault(require("./MultiplePicker"));

var _InputItem = _interopRequireDefault(require("./InputItem"));

var _AddressPicker = _interopRequireDefault(require("./AddressPicker"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
