// @alitajs/generate/lib/generate/pages/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

function _react() {
  const data = _interopRequireDefault(require("react"));

  _react = function _react() {
    return data;
  };

  return data;
}

function _path() {
  const data = require("path");

  _path = function _path() {
    return data;
  };

  return data;
}

function _assert() {
  const data = _interopRequireDefault(require("assert"));

  _assert = function _assert() {
    return data;
  };

  return data;
}

function _chalk() {
  const data = _interopRequireDefault(require("chalk"));

  _chalk = function _chalk() {
    return data;
  };

  return data;
}

function _utils() {
  const data = require("@umijs/utils");

  _utils = function _utils() {
    return data;
  };

  return data;
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance"); }

function _iterableToArrayLimit(arr, i) { if (!(Symbol.iterator in Object(arr) || Object.prototype.toString.call(arr) === "[object Arguments]")) { return; } var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _default({
  api
}) {
  const userConfig = api.userConfig;
  return class PageGenerator extends _utils().Generator {
    constructor(opts) {
      super(opts);
      (0, _assert().default)(typeof this.args._[0] === 'string', `
${_chalk().default.underline.cyan('name')} should be supplied

Example:

  alita g pages users
        `.trim());

      if (userConfig.routes) {
        console.warn(`You should config the routes in config.routes manunally since ${_chalk().default.red('config.routes')} exists`);
        console.log();
      }
    } // prompting() {
    //   if (config.useModel) {
    //     console.warn('useModel,auto create useModel demo');
    //     console.log();
    //   }
    //   const prompts = [
    //     {
    //       name: 'autoCreateModel',
    //       type: 'confirm',
    //       message:
    //         'Do you want to create dva model? 需要创建对应的dva model吗？',
    //       default: true,
    //     },
    //   ];
    //   return this.prompt(prompts).then(props => {
    //     this.prompts = props;
    //   });
    // }


    writing() {
      var _this = this;

      return _asyncToGenerator(function* () {
        const _this$args$_ = _slicedToArray(_this.args._, 1),
              path = _this$args$_[0]; // const jsExt = this.args.typescript ? '.tsx' : '.js';
        // const cssExt = this.args.less ? '.less' : '.css';
        // const path = this.args[0].toString();


        const isTypeScript = true;
        const jsxExt = isTypeScript ? 'tsx' : 'js';
        const jsExt = isTypeScript ? 'ts' : 'js';
        const cssExt = 'less';
        const fileName = (0, _path().basename)(path);

        const componentName = _utils().lodash.upperFirst(fileName);

        const context = {
          name: fileName,
          componentName,
          color: (0, _utils().randomColor)(),
          isTypeScript,
          cssExt,
          jsxExt
        };

        _this.copyTpl({
          templatePath: (0, _path().join)(__dirname, 'index.tsx.tpl'),
          target: (0, _path().join)(api.paths.absPagesPath, `${fileName}/index.${jsxExt}`),
          context
        });

        _this.copyTpl({
          templatePath: (0, _path().join)(__dirname, 'index.less.tpl'),
          target: (0, _path().join)(api.paths.absPagesPath, `${fileName}/index.${cssExt}`),
          context
        });

        _this.copyTpl({
          templatePath: (0, _path().join)(__dirname, 'model.ts.tpl'),
          target: (0, _path().join)(api.paths.absSrcPath, 'models', `${fileName}.${jsExt}`),
          context
        });
      })();
    }

  };
}
