// @alitajs/dform/es/components/MultiplePicker/demo/index.js
import "antd-mobile/es/button/style";
import _Button from "antd-mobile/es/button";
import "antd-mobile/es/white-space/style";
import _WhiteSpace from "antd-mobile/es/white-space";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * title: 基础 多选框
 * desc: 表单使用 demo
 */
import React from 'react';
import DynamicForm, { useForm } from '@alitajs/dform';

var Page = function Page() {
  var _useForm = useForm(),
      _useForm2 = _slicedToArray(_useForm, 1),
      form = _useForm2[0];

  var onFinish = function onFinish(values) {
    // eslint-disable-next-line no-console
    console.log('Success:', values);
  };

  var onFinishFailed = function onFinishFailed(errorInfo) {
    // eslint-disable-next-line no-console
    console.log('Failed:', errorInfo);
  };

  var foodList = [{
    label: '宫保鸡丁',
    value: '宫保鸡丁'
  }, {
    label: '可乐鸡翅',
    value: '可乐鸡翅'
  }, {
    label: '爆炒虾仁',
    value: '爆炒虾仁'
  }, {
    label: '清蒸小黄鱼',
    value: '清蒸小黄鱼'
  }, {
    label: '红烧肉',
    value: '红烧肉'
  }];
  var formsData = [{
    type: 'multiplePicker',
    fieldProps: 'myFood',
    required: true,
    data: foodList,
    title: '我喜欢的食物',
    labelNumber: 7,
    placeholder: '请选择我喜欢的食物',
    onChange: function onChange(e) {
      // eslint-disable-next-line no-console
      console.log(e);
    }
  }, {
    type: 'multiplePicker',
    fieldProps: 'youFood',
    data: foodList,
    title: '选择你喜欢的食物(不可编辑)',
    disabled: true,
    placeholder: '请选择',
    positionType: 'vertical',
    maxValueLength: 1
  }];
  var formsValues = {
    youFood: ['红烧肉', '清蒸小黄鱼']
  };
  var formProps = {
    onFinish: onFinish,
    onFinishFailed: onFinishFailed,
    data: formsData,
    formsValues: formsValues,
    form: form,
    autoLineFeed: false,
    isDev: true
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DynamicForm, formProps), /*#__PURE__*/React.createElement(_WhiteSpace, {
    size: "sm"
  }), /*#__PURE__*/React.createElement(_Button, {
    type: "primary",
    onClick: function onClick() {
      return form.submit();
    }
  }, "Submit"));
};

export default Page;
