// @alitajs/dform/es/components/NomarPicker/index.js
import "antd-mobile/es/picker/style";
import _Picker from "antd-mobile/es/picker";
import "antd-mobile/es/list/style";
import _List from "antd-mobile/es/list";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import Field from '../Field';
import '../../styles/index.less';

var NomarPicker = function NomarPicker(props) {
  var coverStyle = props.coverStyle,
      title = props.title,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      fieldProps = props.fieldProps,
      rules = props.rules,
      placeholder = props.placeholder,
      _props$data = props.data,
      data = _props$data === void 0 ? [] : _props$data,
      _props$positionType = props.positionType,
      positionType = _props$positionType === void 0 ? 'horizontal' : _props$positionType,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      subTitle = props.subTitle,
      _props$hidden = props.hidden,
      hidden = _props$hidden === void 0 ? false : _props$hidden,
      otherProps = _objectWithoutProperties(props, ["coverStyle", "title", "required", "fieldProps", "rules", "placeholder", "data", "positionType", "hasStar", "subTitle", "hidden"]);

  var isVertical = positionType === 'vertical';
  return /*#__PURE__*/React.createElement(React.Fragment, null, !hidden && /*#__PURE__*/React.createElement(React.Fragment, null, isVertical && /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-vertical-title"
  }, required && hasStar && /*#__PURE__*/React.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/React.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title), subTitle), /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform".concat(isVertical ? '-vertical' : '', "-picker")
  }, /*#__PURE__*/React.createElement(Field, {
    name: fieldProps,
    rules: rules || [{
      required: required,
      message: "\u8BF7\u8F93\u5165".concat(title)
    }]
  }, /*#__PURE__*/React.createElement(_Picker, _extends({
    cascade: false
  }, otherProps, {
    style: coverStyle,
    extra: placeholder,
    data: data,
    title: title
  }), /*#__PURE__*/React.createElement(_List.Item, {
    arrow: "horizontal"
  }, !isVertical && /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-title-content"
  }, required && hasStar && /*#__PURE__*/React.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/React.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title))))))));
};

export default NomarPicker;
