// @alitajs/dform/es/components/NomarSwitch/index.js
import "antd-mobile/es/list/style";
import _List from "antd-mobile/es/list";
import "antd-mobile/es/switch/style";
import _Switch from "antd-mobile/es/switch";

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import Field from '../Field';
import '../../styles/index.less';

var NomarSwitch = function NomarSwitch(props) {
  var coverStyle = props.coverStyle,
      title = props.title,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      fieldProps = props.fieldProps,
      rules = props.rules,
      placeholder = props.placeholder,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      _props$hidden = props.hidden,
      hidden = _props$hidden === void 0 ? false : _props$hidden,
      otherProps = _objectWithoutProperties(props, ["coverStyle", "title", "required", "fieldProps", "rules", "placeholder", "hasStar", "hidden"]);

  return /*#__PURE__*/React.createElement(React.Fragment, null, !hidden && /*#__PURE__*/React.createElement(_List.Item, {
    key: fieldProps,
    style: coverStyle,
    extra: /*#__PURE__*/React.createElement(Field, {
      name: fieldProps,
      valuePropName: "checked",
      rules: rules || [{
        required: required,
        message: "\u8BF7\u8F93\u5165".concat(title)
      }]
    }, /*#__PURE__*/React.createElement(_Switch, otherProps))
  }, /*#__PURE__*/React.createElement("div", {
    className: "alitajs-dform-title-content"
  }, required && hasStar && /*#__PURE__*/React.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/React.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title))));
};

export default NomarSwitch;
