// @alitajs/dform/es/components/ExtraInput/demo/index.js
import "antd-mobile/es/button/style";
import _Button from "antd-mobile/es/button";
import "antd-mobile/es/white-space/style";
import _WhiteSpace from "antd-mobile/es/white-space";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

/**
 * title: 基础 复杂输入框
 * desc: 表单使用 demo
 */
import React from 'react';
import DynamicForm, { useForm } from '@alitajs/dform';
import PositionIcon from '../../../assets/position_ico.png';

var Page = function Page() {
  var _useForm = useForm(),
      _useForm2 = _slicedToArray(_useForm, 1),
      form = _useForm2[0];

  var onFinish = function onFinish(values) {
    // eslint-disable-next-line no-console
    console.log('Success:', values);
  };

  var onFinishFailed = function onFinishFailed(errorInfo) {
    // eslint-disable-next-line no-console
    console.log('Failed:', errorInfo);
  };

  var extraImg = function extraImg() {
    return /*#__PURE__*/React.createElement("img", {
      src: PositionIcon
    });
  };

  var unitList = [[{
    label: '元',
    value: '元'
  }, {
    label: '万元',
    value: '万元'
  }, {
    label: '亿元',
    value: '亿元'
  }]];
  var formsData = [{
    type: 'extraInput',
    fieldProps: 'minPrise',
    fieldProps2: 'maxPrise',
    title: '价格区间(数字输入)',
    placeholder: '输入最小价格',
    placeholder2: '输入最大价格',
    required: true,
    inputType: 'number',
    labelNumber: 8,
    extra: '¥'
  }, {
    type: 'extraInput',
    fieldProps: 'minPosition',
    fieldProps2: 'maxPosition',
    title: '位置区间',
    placeholder: '选择最小位置',
    placeholder2: '选择最大位置',
    required: true,
    inputType: 'text',
    extra: extraImg(),
    positionType: 'horizontal'
  }, {
    type: 'extraInput',
    fieldProps: 'minLength',
    fieldProps2: 'maxLength',
    title: '长度区间',
    placeholder: '输入长度',
    placeholder2: '输入长度',
    required: true
  }, {
    type: 'extraInput',
    fieldProps: 'price',
    fieldProps2: 'unit',
    title: '单价',
    placeholder: '输入价格',
    placeholder2: '选择区间',
    required: true,
    extraType: 'select',
    data: unitList,
    positionType: 'horizontal'
  }, {
    type: 'extraInput',
    title: '价格',
    fieldProps: 'prices',
    fieldProps2: 'priceUnit',
    placeholder: '输入长度',
    placeholder2: '选择长度单位',
    positionType: 'vertical',
    extraType: 'select',
    data: unitList,
    firstProps: {
      onChange: function onChange(val) {
        // eslint-disable-next-line no-console
        console.log(val);
      }
    },
    secondProps: {
      onOk: function onOk(val) {
        // eslint-disable-next-line no-console
        console.log(val);
      }
    }
  }];
  var formsValues = {};
  var formProps = {
    onFinish: onFinish,
    onFinishFailed: onFinishFailed,
    data: formsData,
    formsValues: formsValues,
    form: form,
    isDev: true
  };
  return /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(DynamicForm, formProps), /*#__PURE__*/React.createElement(_WhiteSpace, {
    size: "sm"
  }), /*#__PURE__*/React.createElement(_Button, {
    type: "primary",
    onClick: function onClick() {
      return form.submit();
    }
  }, "Submit"));
};

export default Page;
