// @alitajs/dform/es/components/ExtraInput/demo/positionInput.js
import "antd-mobile/es/list/style";
import _List from "antd-mobile/es/list";

/**
 * title: 基础 复杂输入框
 * desc: 位置输入框
 */
import React from 'react';
import Form from 'rc-field-form';
import PositionIcon from '../../../assets/position_ico.png';
import ExtraInput from '..'; // eslint-disable-next-line no-console

var extraImg = function extraImg() {
  return /*#__PURE__*/React.createElement("img", {
    src: PositionIcon,
    onClick: function onClick(e) {
      return console.log(e);
    }
  });
};

var PositionInput = function PositionInput() {
  return /*#__PURE__*/React.createElement(Form, null, /*#__PURE__*/React.createElement(_List, null, /*#__PURE__*/React.createElement(ExtraInput, {
    fieldProps: "minPosition",
    fieldProps2: "maxPosition",
    title: "\u4F4D\u7F6E\u533A\u95F4",
    placeholder: "\u9009\u62E9\u6700\u5C0F\u4F4D\u7F6E",
    placeholder2: "\u9009\u62E9\u6700\u5927\u4F4D\u7F6E",
    required: true,
    inputType: "text",
    extra: extraImg()
  })));
};

export default PositionInput;
