// @alitajs/dform/lib/components/NomarSwitch/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

require("antd-mobile/es/switch/style");

var _switch = _interopRequireDefault(require("antd-mobile/es/switch"));

var _react = _interopRequireDefault(require("react"));

var _Field = _interopRequireDefault(require("../Field"));

require("../../styles/index.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var NomarSwitch = function NomarSwitch(props) {
  var coverStyle = props.coverStyle,
      title = props.title,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      fieldProps = props.fieldProps,
      rules = props.rules,
      placeholder = props.placeholder,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      _props$hidden = props.hidden,
      hidden = _props$hidden === void 0 ? false : _props$hidden,
      otherProps = _objectWithoutProperties(props, ["coverStyle", "title", "required", "fieldProps", "rules", "placeholder", "hasStar", "hidden"]);

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !hidden && /*#__PURE__*/_react.default.createElement(_list.default.Item, {
    key: fieldProps,
    style: coverStyle,
    extra: /*#__PURE__*/_react.default.createElement(_Field.default, {
      name: fieldProps,
      valuePropName: "checked",
      rules: rules || [{
        required: required,
        message: "\u8BF7\u8F93\u5165".concat(title)
      }]
    }, /*#__PURE__*/_react.default.createElement(_switch.default, otherProps))
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-title-content"
  }, required && hasStar && /*#__PURE__*/_react.default.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title))));
};

var _default = NomarSwitch;
exports.default = _default;
