// @alitajs/dform/lib/components/RangeDatePicker/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd-mobile/es/date-picker/style");

var _datePicker = _interopRequireDefault(require("antd-mobile/es/date-picker"));

require("antd-mobile/es/list/style");

var _list = _interopRequireDefault(require("antd-mobile/es/list"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Field = _interopRequireDefault(require("../Field"));

var _utils = require("../../utils");

require("../../styles/index.less");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var RangeDatePicker = function RangeDatePicker(props) {
  var _useState = (0, _react.useState)(),
      _useState2 = _slicedToArray(_useState, 2),
      beginDate = _useState2[0],
      setBeginDate = _useState2[1];

  var _useState3 = (0, _react.useState)(),
      _useState4 = _slicedToArray(_useState3, 2),
      endDate = _useState4[0],
      setEndDate = _useState4[1];

  var fieldProps = props.fieldProps,
      fieldProps2 = props.fieldProps2,
      _props$placeholder = props.placeholder,
      placeholder = _props$placeholder === void 0 ? '开始时间' : _props$placeholder,
      _props$placeholder2 = props.placeholder2,
      placeholder2 = _props$placeholder2 === void 0 ? '结束时间' : _props$placeholder2,
      _props$required = props.required,
      required = _props$required === void 0 ? false : _props$required,
      _props$modeType = props.modeType,
      modeType = _props$modeType === void 0 ? 'date' : _props$modeType,
      _props$rules = props.rules,
      rules = _props$rules === void 0 ? [] : _props$rules,
      title = props.title,
      minDate = props.minDate,
      maxDate = props.maxDate,
      _props$positionType = props.positionType,
      positionType = _props$positionType === void 0 ? 'vertical' : _props$positionType,
      _props$hasStar = props.hasStar,
      hasStar = _props$hasStar === void 0 ? true : _props$hasStar,
      secondProps = props.secondProps,
      firstProps = props.firstProps,
      subTitle = props.subTitle,
      _props$hidden = props.hidden,
      hidden = _props$hidden === void 0 ? false : _props$hidden,
      otherProps = _objectWithoutProperties(props, ["fieldProps", "fieldProps2", "placeholder", "placeholder2", "required", "modeType", "rules", "title", "minDate", "maxDate", "positionType", "hasStar", "secondProps", "firstProps", "subTitle", "hidden"]);

  var isVertical = positionType === 'vertical';
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !hidden && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, isVertical && /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-vertical-title"
  }, required && hasStar && /*#__PURE__*/_react.default.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title), subTitle), /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)({
      'alitajs-dform-range-horizontal': !isVertical,
      'alitajs-dform-range-date-picker': true
    }),
    style: {
      justifyContent: isVertical ? 'space-between' : 'center',
      width: isVertical ? '' : '100%'
    }
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-begin".concat(isVertical ? '-vertical' : '', "-picker")
  }, /*#__PURE__*/_react.default.createElement(_Field.default, {
    name: fieldProps,
    rules: rules || [{
      required: required,
      message: "\u8BF7\u9009\u62E9".concat(title)
    }],
    shouldUpdate: function shouldUpdate(prevValue, nextValue) {
      setBeginDate(nextValue && nextValue[fieldProps]);
      return prevValue !== nextValue;
    }
  }, /*#__PURE__*/_react.default.createElement(_datePicker.default, _extends({}, otherProps, firstProps, {
    mode: modeType,
    extra: placeholder,
    minDate: minDate,
    maxDate: endDate || maxDate,
    title: title,
    format: function format(value) {
      return (0, _utils.changeDateFormat)(value, modeType);
    },
    onChange: function onChange(e) {
      setBeginDate(e);
    }
  }), /*#__PURE__*/_react.default.createElement(_list.default.Item, null, required && hasStar && /*#__PURE__*/_react.default.createElement("span", {
    className: "alitajs-dform-redStar"
  }, "*"), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps,
    className: "alitajs-dform-title"
  }, title), /*#__PURE__*/_react.default.createElement("span", {
    id: fieldProps2
  }))))), /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-line"
  }, "~"), /*#__PURE__*/_react.default.createElement("div", {
    className: "alitajs-dform-end".concat(isVertical ? '-vertical' : '', "-picker")
  }, /*#__PURE__*/_react.default.createElement(_Field.default, {
    name: fieldProps2,
    rules: rules || [{
      required: required,
      message: "\u8BF7\u9009\u62E9".concat(title)
    }],
    shouldUpdate: function shouldUpdate(prevValue, nextValue) {
      setEndDate(nextValue && nextValue[fieldProps2]);
      return prevValue !== nextValue;
    }
  }, /*#__PURE__*/_react.default.createElement(_datePicker.default, _extends({}, otherProps, secondProps, {
    extra: placeholder2,
    mode: modeType,
    minDate: beginDate || minDate,
    maxDate: maxDate,
    title: title,
    format: function format(value) {
      return (0, _utils.changeDateFormat)(value, modeType);
    },
    onChange: function onChange(e) {
      setEndDate(e);
    }
  }), /*#__PURE__*/_react.default.createElement(_list.default.Item, {
    arrow: "horizontal"
  })))))));
};

var _default = RangeDatePicker;
exports.default = _default;
